### PR TITLE
fix issue: DAG's on_failure_callback is not invoked when task failed during testing dag.

### PR DESCRIPTION
**Pull request by mis98zb**
The DAG.test() code is as below:
```python
        # Instead of starting a scheduler, we run the minimal loop possible to check
        # for task readiness and dependency management. This is notably faster
        # than creating a BackfillJob and allows us to surface logs to the user
        while dr.state == State.RUNNING:
            schedulable_tis, _ = dr.update_state(session=session)
            for ti in schedulable_tis:
                add_logger_if_needed(ti)
                ti.task = tasks[ti.task_id]
                _run_task(ti, session=session)
        if conn_file_path or variable_file_path:
            # Remove the local variables we have added to the secrets_backend_list
            secrets_backend_list.pop(0)
```
When exception occurred in a task, the exception will be raised from `_run_task(ti, session=session)`.
In the loop, there is no try...except, so the function is terminated immediately.
This is the reason why DAG's on_failure_callback cannot be invoked during testing.

### CODE REVIEW

**Review by potiuk — CHANGES_REQUESTED** [review]
Could you also please add some actual test running this scenario (those tests are in "models/test_dag.py".

**Comment by mis98zb** [comment]
got it.

**Comment by mis98zb** [comment]
Sorry for late. I have added the actual test for dag.test() with failure.

**Comment by mis98zb** [comment]
I have removed the wrong test code. Can anyone help to approve the testing?
thanks & best regards.

File: airflow/models/dag.py [hunk 1]
```diff
@@ -2698,10 +2698,13 @@ def add_logger_if_needed(ti: TaskInstance):
...
-                add_logger_if_needed(ti)
-                ti.task = tasks[ti.task_id]
-                _run_task(ti, session=session)
+            try:
+                for ti in schedulable_tis:
+                    add_logger_if_needed(ti)
+                    ti.task = tasks[ti.task_id]
+                    _run_task(ti, session=session)
+            except Exception:
+                self.log.info("Task failed, continue looping. DAG will be marked as failed at the end of looping.")
```
Comment (uranusjr) [inline airflow/models/dag.py]:
```suggestion
                self.log.info(
                    "Task %s failed. DAG will continue to run until finished "
                    "and be marked as failed.",
                    ti.key,
                )
```

Comment (uranusjr) [reply airflow/models/dag.py]: This wording is likely better than “looping”.

File: airflow/models/dag.py [hunk 2]
```diff
@@ -2698,10 +2698,16 @@ def add_logger_if_needed(ti: TaskInstance):
...
+            try:
+                for ti in schedulable_tis:
+                    add_logger_if_needed(ti)
+                    ti.task = tasks[ti.task_id]
+                    _run_task(ti, session=session)
+            except Exception:
+                self.log.info(
+                    "Task failed. DAG will continue to run until finished "
+                    "and be marked as failed."
+                )
```
Comment (uranusjr) [inline airflow/models/dag.py]: Should we add the exception to the log so the user knows what happened? Or would that be unhelpful? (Not sure.)

Comment (uranusjr) [inline airflow/models/dag.py]:
```suggestion
                self.log.info(
                    "Task failed. DAG will continue to run until finished "
                    "and be marked as failed.",
                    exc_info=True,
                )
```

If we want to include the exception.

**Review by uranusjr — APPROVED** [review]

File: airflow/models/dag.py (hunk 2 above)
Comment (mis98zb) [reply airflow/models/dag.py]: I see. Thank you for the advise.

**Comment by mis98zb** [comment]
@potiuk I have added some tests for the code change. May I know if the is enough? Thanks & Best Regards.

**Review by uranusjr — APPROVED** [review]

**Review by potiuk — APPROVED** [review]

**Comment by mis98zb** [comment]
Don't know why the build keep failing, ToT
Seems the failed test is not related to the code change in this PR....

**Comment by mis98zb** [comment]
May I know how can I format code in local machine?

I tried `--config .\pyproject.toml airflow`, but it cannot get the same result in the build...

**Comment by uranusjr** [comment]
Please set up pre-commit as described in the contributor’s guide.

**Comment by potiuk** [comment]
> Please set up pre-commit as described in the contributor’s guide.

The instructions were actually printed by pre-commit, but could be lost in see of messages - I made PR to display far more prominent https://github.com/apache/airflow/pull/31813

**Review by potiuk — APPROVED** [review]

**Pull request merged**